### PR TITLE
Refactor Piston to set properties immutably

### PR DESCRIPTION
**Pull request by settermjd**
Whilst this isn't strictly necessary, I find using `setOffsetLimit()`
jarring; primarily because it was not following the convention of
`Zend\Diactoros\ServerRequest`. Given that, I've added a revised
method, `withOffsetLimit()`, which follows that convention.

### CODE REVIEW

**Comment by settermjd** [comment]
Happy to leave this till later on.

File: src/Request.php [hunk 1]
```diff
@@ -32,28 +32,29 @@ public function __construct(
...
     /**
-     * @var null
+     * @var array
      */
-    protected $requestedFields = null;
+    protected $requestedFields;
 
     /**
-     * @var null
+     * @var array
```
Comment (rcatlin) [inline src/Request.php]: The default is `null`. Change `@var array` to `@var array|null`?

File: src/Middleware/Payload.php [hunk 2]
```diff
@@ -52,4 +52,19 @@ public function getResponse()
...
     }
+
+    /**
+     * Set the Payload's request object
+     *
+     * @param Request $request
+     *
+     * @return Payload
+     */
+    public function withRequest(Request $request)
```
Comment (kayladnls) [inline src/Middleware/Payload.php]: Why? There's not a way to set a Request currently, and there shouldn't be...

File: src/Request.php (hunk 1 above)
Comment (kayladnls) [inline src/Request.php]: I would make it empty array.

File: src/Request.php [hunk 3]
```diff
@@ -139,6 +148,8 @@ public function setAfterCursor($after_cursor)
 
     /**
      * @param string $before_cursor
+     *
+     * @deprecated
```
Comment (kayladnls) [inline src/Request.php]: This is not deprecated.

File: src/Request.php [hunk 4]
```diff
@@ -130,6 +137,8 @@ public function isPaginated()
 
     /**
      * @param string $after_cursor
+     *
+     * @deprecated
```
Comment (kayladnls) [inline src/Request.php]: not deprecated.

File: src/Request.php [hunk 5]
```diff
@@ -213,10 +231,99 @@ public function clearCookie($key)
         return $this->withCookieParams($this->cookieJar->all());
     }
 
+    /**
+     * @return \Psr\Http\Message\ServerRequestInterface|ServerRequest
```
Comment (kayladnls) [inline src/Request.php]: I don't think this type hint is correct.

File: src/Middleware/Payload.php (hunk 2 above)
Comment (settermjd) [inline src/Middleware/Payload.php]: Good point. Thanks for pointing that out.

**Comment by settermjd** [comment]
For the latest changes to work, #122 needs to be merged to master - and merged with this branch naturally :-)

**Comment by rcatlin** [comment]
:+1: Looks great to me. Would love to get a final thumbs up from @kayladnls

**Comment by settermjd** [comment]
Definitely!

**Comment by rcatlin** [comment]
I think one the conflicts are resolved we can merge. :+1:

**Comment by settermjd** [comment]
I didn't see the conflicts on the mobile version. Will get that sorted now.

**Comment by localheinz** [comment]
@settermjd 

Conflicts were created after your PR was opened, as a bunch of fixes have been merged!

**Comment by settermjd** [comment]
Thanks @localheinz. working through the fix atm

**Comment by settermjd** [comment]
I was confused initially as I'd done a merge from the canonical master to my master, then merged that in to my branch mid-last week, so thought it must have been an old reference.

**Comment by settermjd** [comment]
@rcatlin @kayladnls @localheinz - ready to merge?

**Comment by localheinz** [comment]
@settermjd 

:+1:

**Comment by rcatlin** [comment]
Thanks @settermjd!

**Pull request merged**